### PR TITLE
Feature FavoriteCategories

### PR DESCRIPTION
**Pull request by jenniferlim**

### CODE REVIEW

**Comment by joelevin** [comment]
this PR includes https://github.com/jenniferlim/weTranslate/pull/24 :baby:

File: weTranslate/FavoriteCategory.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    let language: Language
+    let translations: [Translation]
+
+
+    // MARK: - Initialization
+
+    init?(dictionary: JSONDictionary) {
+        guard let languageRawValue = dictionary["language"] as? String,
+            language = Language(rawValue: languageRawValue),
+        translationDictionaries = dictionary["translations"] as? [JSONDictionary] else { return nil }
```
Comment (lionel-alves) [inline weTranslate/FavoriteCategory.swift]: :arrow_right:

File: weTranslate/FavoriteCategory.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    // MARK: - Properties
+
+    let language: Language
+    let translations: [Translation]
+
+
+    // MARK: - Initialization
+
+    init?(dictionary: JSONDictionary) {
```
Comment (lionel-alves) [inline weTranslate/FavoriteCategory.swift:31]: Implement that in a extension.

File: weTranslateTests/FavoriteCategoryStoreTests.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    func testFetchAll() {
+
+        guard let dictionaries = loadJSON("favoriteCategoriesStore") as? [JSONDictionary],
+            favoriteCategory = FavoriteCategory(dictionary: dictionaries[0]),
+            translation = favoriteCategory.translations.first else {
+                XCTFail()
+                return
+        }
+
+
```
Comment (lionel-alves) [inline weTranslateTests/FavoriteCategoryStoreTests.swift]: :scissors:

File: weTranslateTests/FavoriteCategoryStoreTests.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        let favoriteCategoryStore = FavoriteCategoryStore()
+        favoriteCategoryStore.insert(translation: translation)
+        let favoriteCategories = favoriteCategoryStore.fetchAll()
+
+        XCTAssertEqual(favoriteCategories.count, 1)
+        XCTAssertEqual(favoriteCategories[0].translations.count, 1)
+        XCTAssertEqual(favoriteCategories[0].translations[0].meanings[0].translatedWords.count, 2)
+    }
+
```
Comment (lionel-alves) [inline weTranslateTests/FavoriteCategoryStoreTests.swift]: :scissors:

File: Cartfile [hunk 5]
```diff
@@ -1 +1 @@
-github "https://github.com/lionel-alves/TranslateKit" "ebe1f2ded03dfedc448bdac66a78257ca81d4ef2"
+github "https://github.com/lionel-alves/TranslateKit" "jl/translation-init-deserialiazation"
```
Comment (jenniferlim) [inline Cartfile]: Pls update to Develop

File: Cartfile.resolved [hunk 6]
```diff
@@ -1,2 +1,2 @@
 github "venmo/DVR" "v0.2.1"
-github "lionel-alves/TranslateKit" "ebe1f2ded03dfedc448bdac66a78257ca81d4ef2"
+github "lionel-alves/TranslateKit" "63426d245a8522b693ed848de578c4bb602e29f5"
```
Comment (jenniferlim) [inline Cartfile.resolved]: Pls update to Develop

File: weTranslate/FavoriteCategoriesTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        }
+    }
+
+    private let overlayView: UIView = {
+        let view = UIView(frame: .zero)
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = .blackColor()
+        view.layer.opacity = 0.2
+        return view
+    }()
```
Comment (jenniferlim) [inline weTranslate/FavoriteCategoriesTableViewCell.swift:31]: Could we create a generic UIView like what we did for Label and TableViewCell

File: weTranslate/FavoriteCategoriesTableViewCell.swift [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    private let languageLabel: Label = {
+        let label = Label(textColor: .whiteColor(), font: Font.font(style: .Title1, weight: .Medium))
+        label.numberOfLines = 0
+        return label
+    }()
+
+    private let columnView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.translatesAutoresizingMaskIntoConstraints = false
```
Comment (jenniferlim) [inline weTranslate/FavoriteCategoriesTableViewCell.swift:63]: Could we add a generic StackView also

File: weTranslate/FavoriteCategoriesViewController.swift [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+final class FavoriteCategoriesViewController: UIViewController {
+
+    // MARK: - Properties
+
+    var viewModel: [FavoriteCategoryViewModel]? {
+        didSet {
+            dispatch_async(dispatch_get_main_queue()) { [weak self] () -> Void in
+                self?.tableView.reloadData()
```
Comment (jenniferlim) [inline weTranslate/FavoriteCategoriesViewController.swift]: Could we one line this code

File: weTranslate/FavoriteCategoriesTableViewCell.swift (hunk 7 above)
Comment (lionel-alves) [inline weTranslate/FavoriteCategoriesTableViewCell.swift:31]: https://github.com/jenniferlim/weTranslate/issues/38

File: weTranslate/FavoriteCategoriesTableViewCell.swift (hunk 8 above)
Comment (lionel-alves) [inline weTranslate/FavoriteCategoriesTableViewCell.swift:63]: https://github.com/jenniferlim/weTranslate/issues/38

File: weTranslate/FavoriteCategoriesTableViewCell.swift [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+        contentView.addSubview(overlayView)
+        contentView.addSubview(columnView)
+        backgroundView = backgroundImageView
+
+        let margins = contentView.layoutMarginsGuide
+        columnView.leadingAnchor.constraintEqualToAnchor(margins.leadingAnchor).active = true
+        columnView.trailingAnchor.constraintEqualToAnchor(margins.trailingAnchor).active = true
+        columnView.topAnchor.constraintEqualToAnchor(margins.topAnchor).active = true
+        columnView.bottomAnchor.constraintEqualToAnchor(margins.bottomAnchor).active = true
```
Comment (jenniferlim) [inline weTranslate/FavoriteCategoriesTableViewCell.swift:102]: We might want a category on UIView that take a view (layoutGuide or view) and apply those constraints.

**Comment by jenniferlim** [comment]
:bowtie:

File: weTranslate/FavoriteCategoriesTableViewCell.swift (hunk 10 above)
Comment (lionel-alves) [inline weTranslate/FavoriteCategoriesTableViewCell.swift:102]: https://github.com/jenniferlim/weTranslate/issues/39

**Pull request merged**